dotbot/protocol.py

```diff
@@ -59,543 +62,485 @@ class ProtocolPayloadParserException(Exception):
-                        byteorder="little",
-                    )
-                )
-        waypoints = [
-            (waypoints[i], waypoints[i + 1], waypoints[i + 2])
-            for i in range(0, 3 * waypoints_count, 3)
+    metadata: list[PacketFieldMetadata] = dataclasses.field(
+        default_factory=lambda: [
+            PacketFieldMetadata(name="threshold", disp="thr."),
+            PacketFieldMetadata(name="count", disp="len."),
+            PacketFieldMetadata(name="waypoints", type_=list, length=0),
         ]
-        return LH2Waypoints(threshold=threshold, waypoints=waypoints)
+    )
 
+    threshold: int = 0
+    count: int = 0
+    waypoints: list[PayloadGPSPosition] = dataclasses.field(default_factory=lambda: [])
 
-@dataclass
-class GPSWaypoints(ProtocolData):
-    """Dataclass that holds a list of GPS waypoints."""
 
-    threshold: int
-    waypoints: List[GPSPosition] = dataclasses.field(default_factory=lambda: [])
+PAYLOAD_PARSERS: dict[PayloadType, Packet] = {
+    PayloadType.ADVERTISEMENT: PayloadAdvertisement,
+    PayloadType.CMD_MOVE_RAW: PayloadCommandMoveRaw,
+    PayloadType.CMD_RGB_LED: PayloadCommandRgbLed,
+    PayloadType.CMD_XGO_ACTION: PayloadCommandXgoAction,
+    PayloadType.LH2_RAW_LOCATION: PayloadLh2RawLocation,
+    PayloadType.LH2_PROCESSED_DATA: PayloadLh2ProcessedLocation,
+    PayloadType.LH2_RAW_DATA: PayloadLh2RawData,
+    PayloadType.LH2_LOCATION: PayloadLH2Location,
+    PayloadType.DOTBOT_DATA: PayloadDotBotData,
+    PayloadType.GPS_POSITION: PayloadGPSPosition,
+    PayloadType.SAILBOT_DATA: PayloadSailBotData,
+    PayloadType.DOTBOT_SIMULATOR_DATA: PayloadDotBotSimulatorData,
+    PayloadType.CONTROL_MODE: PayloadControlMode,
+    PayloadType.LH2_WAYPOINTS: PayloadLH2Waypoints,
+    PayloadType.GPS_WAYPOINTS: PayloadGPSWaypoints,
+}
 
-    @property
-    def fields(self) -> List[ProtocolField]:
-        _fields = [ProtocolField(len(self.waypoints), name="len.")]
-        _fields += [ProtocolField(value=self.threshold, name="thr.")]
-        _fields += list(chain(*[waypoint.fields for waypoint in self.waypoints]))
-        return _fields
-
-    @staticmethod
-    def from_bytes(bytes_) -> ProtocolData:
-        waypoints_count = int(bytes_[0])
-        threshold = int(bytes_[1])
-        waypoints_bytes = bytes_[2:]
-        waypoints = []
-        for idx in range(2 * waypoints_count):
-            waypoints.append(
-                float(
-                    int.from_bytes(
-                        waypoints_bytes[4 * idx : 4 * (idx + 1)], byteorder="little"
-                    )
-                    / 1e6
-                )
-            )
-        waypoints = [
-            (waypoints[i], waypoints[i + 1]) for i in range(0, 2 * waypoints_count, 2)
-        ]
-        return GPSWaypoints(threshold=threshold, waypoints=waypoints)
 
+def register_parser(payload_type: PayloadType, parser):
+    PAYLOAD_PARSERS[payload_type] = parser
 
-@dataclass
-class ProtocolPayload:
```

Comment:
Really cool what you did for automagically parsing to/from bytes using only the descriptions in the dataclass itself!